codalab/lib/beam/blobstorageuploader.py

```diff
@@ -0,0 +1,99 @@
+from apache_beam.io.filesystemio import Uploader, UploaderStream
+from azure.storage.blob import (
+    BlobServiceClient,
+    ContentSettings,
+    BlobBlock,
+)
+from apache_beam.io.azure.blobstorageio import parse_azfs_path
+import tempfile
+import io
+import base64
+from codalab.worker.un_gzip_stream import BytesBuffer
+
+class BlobStorageUploader(Uploader):
+  """An improved version of apache_beam.io.azure.blobstorageio.BlobStorageUploader
+  that handles multipart streaming (block-by-block) uploads.
+  TODO (Ashwin): contribute this back upstream to Apache Beam (https://github.com/codalab/codalab-worksheets/issues/3475).
+  """
+  def __init__(self, client, path, mime_type='application/octet-stream'):
+    self._client = client
+    self._path = path
+    self._container, self._blob = parse_azfs_path(path)
+    self._content_settings = ContentSettings(mime_type)
+
+    self._blob_to_upload = self._client.get_blob_client(
+        self._container, self._blob)
+
+    self.block_number = 1
+    self.buffer = BytesBuffer()
+    self.block_list = []
+
+  def put(self, data):
+    # Note that Blob Storage currently can hold a maximum of 100,000 uncommitted blocks.
+    # This means that with this current implementation, we can upload a file with a maximum
+    # size of 10 TiB to Blob Storage. To exceed that limit, we must either increase MIN_WRITE_SIZE
+    # or modify the implementation of this class to call commit_block_list more often (and not
+    # just at the end of the upload). 
+    MIN_WRITE_SIZE = 100 * 1024 * 1024
+    # Maximum block size is 4000 MiB (https://docs.microsoft.com/en-us/rest/api/storageservices/put-block#remarks).
```

Comment:
@adiprerepa FYI I changed the max write size to 4000 MiB because that's the max size blob storage supports.